components/calendar/src/chinese_based.rs

```diff
@@ -232,63 +286,133 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
     /// from the fixed date of the beginning of the month.
     ///
     /// The calculation for `elapsed_years` and `month` in this function are based on code from _Calendrical Calculations_ by Reingold & Dershowitz.
     /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5414-L5459
-    pub(crate) fn chinese_based_date_from_fixed(date: RataDie) -> ChineseBasedDateInner<C> {
-        let (first_day_of_year, next_solstice) = Self::new_year_on_or_before_fixed_date(date, None);
-        let next_new_year =
-            Self::new_year_on_or_before_fixed_date(first_day_of_year + 400, Some(next_solstice)).0;
-        let year_float = libm::floor(
-            1.5 - 1.0 / 12.0 + ((first_day_of_year - C::EPOCH) as f64) / MEAN_TROPICAL_YEAR,
-        );
-        let year_int = i64_to_i32(year_float as i64);
-        debug_assert!(
-            matches!(year_int, I32Result::WithinRange(_)),
-            "Year should be in range of i32"
-        );
-        let year = year_int.saturate();
-        let new_moon = Self::new_moon_before((date + 1).as_moment());
-        let month_i64 =
-            libm::round((new_moon - first_day_of_year) as f64 / MEAN_SYNODIC_MONTH) as i64 + 1;
-        debug_assert!(
-            ((u8::MIN as i64)..=(u8::MAX as i64)).contains(&month_i64),
-            "Month should be in range of u8! Value {month_i64} failed for RD {date:?}"
-        );
-        let month = month_i64 as u8;
-        let day_i64 = date - new_moon + 1;
-        debug_assert!(
-            ((u8::MIN as i64)..=(u8::MAX as i64)).contains(&month_i64),
-            "Day should be in range of u8! Value {month_i64} failed for RD {date:?}"
-        );
-        let day = day_i64 as u8;
-        let is_leap_year = Self::new_year_is_leap_year(
-            first_day_of_year,
-            Some(next_solstice),
-            Some(next_new_year),
-        );
-        let leap_month = if is_leap_year {
-            // This doesn't need to be checked for `None`, since `get_leap_month_from_new_year`
-            // will always return a number greater than or equal to 1, and less than 14.
-            NonZeroU8::new(Self::get_leap_month_from_new_year(first_day_of_year))
+    pub(crate) fn chinese_based_date_from_fixed(
+        date: RataDie,
+        iso: Option<Date<Iso>>,
+    ) -> ChineseBasedDateInner<C> {
+        let iso = if let Some(iso_date) = iso {
+            iso_date
         } else {
-            None
+            Iso::iso_from_fixed(date)
         };
-        let cache = ChineseBasedCache {
-            new_year: first_day_of_year,
-            next_new_year,
-            leap_month,
+
+        let epoch_as_iso = Iso::iso_from_fixed(C::EPOCH);
+        let year = iso.year().number - epoch_as_iso.year().number + 1;
+
+        let data_option = C::get_compiled_data_for_year(year);
+        let data_option = if let Some(data) = data_option {
```

Comment:
nit: this little dance probably should be a separate function that lets you get compiled data for an RD